### PR TITLE
add unit tests

### LINKED ISSUE #41: Add unit testing
We should start adding basic unit tests and include a GitHub action that runs these.

### PR DESCRIPTION
**Pull request by nojnhuh**
This PR adds unit tests for at least the low-hanging fruit and enables the tests in CI. I think what I didn't include here already has decent coverage from the e2e tests, but please let me know if I missed something that would be valuable to add unit tests for.

Fixes #41

### CODE REVIEW

File: api/example.com/resource/gpu/v1alpha1/api_test.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+				if err == nil {
+					t.Errorf("expected error %q, but got no error", test.expectedErr)
+				} else if err.Error() != test.expectedErr {
+					t.Errorf("expected error %q, but got %v", test.expectedErr, err)
+				}
+			} else {
+				if err != nil {
+					t.Errorf("expected Normalize to succeed, got error %v", err)
+				}
+			}
```
Comment (elezar) [inline api/example.com/resource/gpu/v1alpha1/api_test.go:110]: This can get quite verbose if repeated for each test. @pohly is there a recommended way to handle errors in tests in other k8s projects?

File: api/example.com/resource/gpu/v1alpha1/sharing_test.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	}{
+		{
+			name:        "nil GpuSharing",
+			gpuSharing:  nil,
+			expectedErr: "no sharing set to get config from",
+		},
+		{
+			name: "strategy is not TimeSlicing",
+			gpuSharing: &GpuSharing{
+				Strategy: "not" + TimeSlicingStrategy,
```
Comment (elezar) [inline api/example.com/resource/gpu/v1alpha1/sharing_test.go:40]: Question: Does using a valid non-TimeSlicingStrategy here make sense?

File: api/example.com/resource/gpu/v1alpha1/sharing_test.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	}{
+		{
+			name:        "nil GpuSharing",
+			gpuSharing:  nil,
+			expectedErr: "no sharing set to get config from",
+		},
+		{
+			name: "strategy is not SpacePartitioning",
+			gpuSharing: &GpuSharing{
+				Strategy: "not" + SpacePartitioningStrategy,
```
Comment (elezar) [inline api/example.com/resource/gpu/v1alpha1/sharing_test.go:102]: similar comment as above.

File: api/example.com/resource/gpu/v1alpha1/validate_test.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+import (
+	"testing"
+)
+
+func TestGpuConfigValidate(t *testing.T) {
```
Comment (elezar) [inline api/example.com/resource/gpu/v1alpha1/validate_test.go:23]: One test case that is not covered is that if both TimeSlicing and SpacePartitioning are defined, an invalid SpacePartitioning config is ignored.

This might not be intended behaviour, but it may be behaviour that users start to depend on. (although since this is an example driver it's probably not critical) cc @pohly @klueska

**Review by elezar — COMMENTED** [review]
Thanks @nojnhuh. I've made a first pass.

I have a few minor questions and comments.

File: api/example.com/resource/gpu/v1alpha1/api_test.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
```
[the remark this answers is not in the document]
Comment (pohly) [reply api/example.com/resource/gpu/v1alpha1/api_test.go:22]: I find assertions with github.com/stretchr/testify easier to read and write, but that might just be me.

[the remark this answers is not in the document]
Comment (nojnhuh) [reply api/example.com/resource/gpu/v1alpha1/api_test.go:22]: I like using cmp for the diff you get in the failure message, but I see testify's `Equal` also has that built in, so I can switch these over to that.

File: api/example.com/resource/gpu/v1alpha1/api_test.go (hunk 1 above)
Comment (nojnhuh) [reply api/example.com/resource/gpu/v1alpha1/api_test.go:110]: This gets cleaned up a fair bit using testify's `ErrorEqual` and `NoError`, but if there's a premade wrapper around those together that should work here. I could make that kind of wrapper here too.

File: api/example.com/resource/gpu/v1alpha1/sharing_test.go (hunk 2 above)
Comment (nojnhuh) [reply api/example.com/resource/gpu/v1alpha1/sharing_test.go:40]: Yeah that would be equivalent here. I tend toward this kind of mangling of the "correct" value make it "incorrect" to guard against the extreme edge case where two constants like `TimeSlicingStrategy` and `SpacePartitioningStrategy` actually have the same value. Overall though I just want to make these tests as least surprising or interesting as possible, so I can change this.

File: api/example.com/resource/gpu/v1alpha1/validate_test.go (hunk 4 above)
Comment (nojnhuh) [reply api/example.com/resource/gpu/v1alpha1/validate_test.go:23]: Agree, that seems worth recording in a test, even if just to signal to authors of real drivers that that's something to look out for and potentially not recreate.

**Comment by nojnhuh** [comment]
Pushed a merge commit to fix the conflict in case that's easier to review than a rebase, but I'll plan to squash before merging this.

/hold

File: api/example.com/resource/gpu/v1alpha1/api_test.go (hunk 1 above)
Comment (elezar) [reply api/example.com/resource/gpu/v1alpha1/api_test.go]: let's just start with:
```
if test.expecteError != "" {
   require.EqualError(t, err, test.expecteError)
} else {
   require.NoError(t, err)
}
```
We can look at a more comprehensive wrapper later if required.

File: go.mod [hunk 6]
```diff
@@ -3,8 +3,10 @@ module sigs.k8s.io/dra-example-driver
 go 1.23.1
 
 require (
+	github.com/google/go-cmp v0.6.0
```
Comment (elezar) [inline go.mod:6]: I thought we removed the use of `cmp`?

**Review by elezar — COMMENTED** [review]
Minor comment on the go modules, but otherwise this looks good.

File: api/example.com/resource/gpu/v1alpha1/api_test.go (hunk 1 above)
Comment (elezar) [reply api/example.com/resource/gpu/v1alpha1/api_test.go]: (I see that's what you've already done!).

File: api/example.com/resource/gpu/v1alpha1/api_test.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2024 The Kubernetes Authors.
```
Comment (bart0sh) [inline api/example.com/resource/gpu/v1alpha1/api_test.go:2]: 2025?

File: api/example.com/resource/gpu/v1alpha1/api_test.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+package v1alpha1
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGpuConfigNormalize(t *testing.T) {
+	tests := []struct {
```
Comment (bart0sh) [inline api/example.com/resource/gpu/v1alpha1/api_test.go:26]: nit: Making it a map would change an order of test cases for each run, which can help ensuring that they don't influence each other.
```suggestion
	tests := map[string]struct {
```

File: go.mod (hunk 6 above)
Comment (nojnhuh) [reply go.mod]: Ah yep, just missed a `go mod tidy`.

File: api/example.com/resource/gpu/v1alpha1/api_test.go (hunk 7 above)
Comment (nojnhuh) [reply api/example.com/resource/gpu/v1alpha1/api_test.go]: Happy New Year!

**Comment by nojnhuh** [comment]
I took the liberty of squashing after addressing the latest round of feedback.

/hold cancel

File: api/example.com/resource/gpu/v1alpha1/validate_test.go [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGpuConfigValidate(t *testing.T) {
+	tests := map[string]struct {
+		gpuConfig *GpuConfig
+		expected  string
```
Comment (bart0sh) [inline api/example.com/resource/gpu/v1alpha1/validate_test.go:28]: nit: changing type to `error` would simplify the code a little bit and make it more logical in my opinion as `expected` is what `gpuConfig.Validate()` returns.
```suggestion
		expected  error
```

File: api/example.com/resource/gpu/v1alpha1/validate_test.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+		"empty GpuConfig": {
+			gpuConfig: &GpuConfig{},
+			expected:  "no sharing strategy set",
+		},
+		"empty GpuConfig.Sharing": {
+			gpuConfig: &GpuConfig{
+				Sharing: &GpuSharing{},
+			},
+			expected: "unknown GPU sharing strategy: ",
+		},
```
Comment (bart0sh) [inline api/example.com/resource/gpu/v1alpha1/validate_test.go:39]: I'd also add this test case:
```
		"unknown GPU sharing strategy": {
			gpuConfig: &GpuConfig{
				Sharing: &GpuSharing{
					Strategy: "unknown",
				},
			},
			expected: errors.New("unknown GPU sharing strategy: unknown"),
		},
```
BTW, this test case shows that[ this code](https://github.com/kubernetes-sigs/dra-example-driver/blob/main/api/example.com/resource/gpu/v1alpha1/validate.go#L65) is unreachable.

**Review by elezar — APPROVED** [review]
Lgtm
